server-spi-private/src/main/java/org/keycloak/models/locking/GlobalLockProvider.java

```diff
@@ -17,57 +17,73 @@
+     *                                       duration is determined)
+     * @throws NullPointerException          When lockName is {@code null}.
      */
-    default GlobalLock acquireLock(String lockName) throws LockAcquiringTimeoutException {
-        return acquire(lockName, null);
+    default <V> V withLock(String lockName, KeycloakSessionTaskWithResult<V> task) throws LockAcquiringTimeoutException {
+        return withLock(lockName, null, task);
     }
 
     /**
-     * Acquires a new global lock that is visible to all Keycloak nodes. The lock is non-reentrant.
-     * <p />
-     * The lock is guaranteed to be kept until the returned {@link GlobalLock} is closed
-     * using the {@link GlobalLock#close} method.
-     * <p />
-     * Some implementations may benefit from locks that are released at the end of transaction.
-     * For this purpose, the lifespan of the returned lock is limited by the transaction lifespan
-     * of the session which acquired this lock.
-     * <p />
-     * This method is intended to be used in a {@code try}-with-resources block.
-     * <p />
+     * Acquires a new non-reentrant global lock that is visible to all Keycloak nodes. If the lock was successfully
+     * acquired the method runs the {@code task} in a new transaction using method
+     * {@link org.keycloak.models.utils.KeycloakModelUtils#runJobInTransactionWithResult(KeycloakSessionFactory, KeycloakSessionTaskWithResult)}.
+     * <p/>
      * If there is another global lock with the same identifier ({@code lockName}) already acquired, this method waits
      * until the lock is released, however, not more than {@code timeToWaitForLock} duration. If the lock is not
-     * acquired after {@code timeToWaitForLock} duration, the method throws {@link LockAcquiringTimeoutException}
-     * <p />
-     * Releasing of the lock is done using instance of {@link GlobalLock} returned by this method.
+     * acquired after {@code timeToWaitForLock} duration, the method throws {@link LockAcquiringTimeoutException}.
+     * <p/>
+     * When the execution of the {@code task} finishes, the acquired lock must be released regardless of the result.
+     * <p/>
+     * To make sure acquiring/releasing the lock is visible to all Keycloak nodes it may be needed to run the code that
+     * acquires/releases the lock in a separate transactions. This means together the method can use 3 separate
+     * transactions, for example:
+     * <pre>
+     *     try {
+     *         KeycloakModelUtils.runJobInTransaction(factory,
+     *                                innerSession -> /* run code that acquires the lock *\/)
+     *
+     *         KeycloakModelUtils.runJobInTransactionWithResult(factory, task)
+     *     } finally {
+     *         KeycloakModelUtils.runJobInTransaction(factory,
+     *                                innerSession -> /* run code that releases the lock *\/)
+     *     }
+     * </pre>
```

Comment:
The JavaDoc switches the addresses personal from a caller to an implementor. As the usual personal is that of a caller, switching the persona to an implementator should be stated clearly.

```suggestion
     * When the execution of the {@code task} finishes, it releases the acquired lock regardless of the result or if an exception has been thrown.
     * <p/>
     * <b>A note to implementors of the interface:</b>
     * <p/>
     * To make sure acquiring/releasing the lock is visible to all Keycloak nodes it may be needed to run the code that
     * acquires/releases the lock in separate transactions. This means together the method can use 3 separate
     * transactions, for example:
     * <pre>
     *     try {
     *         KeycloakModelUtils.runJobInTransaction(factory,
     *                                innerSession -> /* run code that acquires the lock *\/);
     *
     *         KeycloakModelUtils.runJobInTransactionWithResult(factory, task);
     *     } finally {
     *         KeycloakModelUtils.runJobInTransaction(factory,
     *                                innerSession -> /* run code that releases the lock *\/);
     *     }
     * </pre>
```